### ISSUE TITLE
No support for optimistic locks

### ISSUE THREAD

**Issue by ajays20078** [jinzhu/gorm#400]
Don't see how to achieve "select for update" using this orm. Doesn't it support implementation of  locking rows/tableS?

**Comment by jinzhu** [comment]
Hi @ajays20078 

Could you let me know your use case, then maybe I could give some suggestions for how to do that.

**Comment by ajays20078** [comment]
Hi @jinzhu 

Use case:

type Resource struct {
            ID int,
           status string
} where status is either available or allotted.

Now  if there are 2 parallel requests and if i have to select 5 resources and update them as allotted "select * from resource where status=\"available\" limit 5"  . Select statement will return same 5 ids/resources to both requests as they are parallel requests and you are not doing "select for update"

**Comment by jinzhu** [comment]
Are you looking for:

```go
db.Where("state = ?", "available").Limit(5).Update(Resource{Status: "allotted"})
```

**Comment by ajays20078** [comment]
select * from resource;
+----+-----------+
| id | status    |
+----+-----------+
|  1 | available |
+----+-----------+
1 row in set (0.00 sec)


var resource Resource
db.Where("status = ?", "avaliable").Limit(1).Update(Resource{status:
"allotted"}).Find(&resource)


panics:

panic: reflect: call of reflect.Value.MethodByName on zero Value

goroutine 16 [running]:
runtime.panic(0x6715a0, 0xc2080475c0)
/usr/share/go/src/pkg/runtime/panic.c:279 +0xf5
reflect.Value.MethodByName(0x0, 0x0, 0x0, 0x0, 0x6d8390, 0xb, 0x0, 0x0,
0x0, 0x0)
/usr/share/go/src/pkg/reflect/value.go:1306 +0xb1
github.com/jinzhu/gorm.func·005(0x0
<http://github.com/jinzhu/gorm.func%C2%B7005(0x0>, 0x0)
/usr/share/go/bin/src/github.com/jinzhu/gorm/scope.go:174 +0xae
github.com/jinzhu/gorm.(*Scope).CallMethod(0xc208048200, 0x6d8390, 0xb, 0x1)
/usr/share/go/bin/src/github.com/jinzhu/gorm/scope.go:199 +0x2a3
github.com/jinzhu/gorm.(*Scope).CallMethodWithErrorCheck(0xc208048200,
0x6d8390, 0xb)
/usr/share/go/bin/src/github.com/jinzhu/gorm/scope.go:204 +0x40
github.com/jinzhu/gorm.AfterUpdate(0xc208048200)
/usr/share/go/bin/src/github.com/jinzhu/gorm/callback_update.go:70 +0x68
github.com/jinzhu/gorm.(*Scope).callCallbacks(0xc208048200, 0xc208054880,
0x9, 0x10, 0xc2080462c0)
/usr/share/go/bin/src/github.com/jinzhu/gorm/scope_private.go:297 +0x6b
github.com/jinzhu/gorm.(*DB).Updates(0xc208040b00, 0x689500, 0xc2080462c0,
0x7f90d7408d87, 0x1, 0x1, 0x7f90d7408f30)
/usr/share/go/bin/src/github.com/jinzhu/gorm/main.go:250 +0x139
github.com/jinzhu/gorm.(*DB).Update(0xc208040b00, 0x7f90d7408f08, 0x1, 0x1,
0x1)
/usr/share/go/bin/src/github.com/jinzhu/gorm/main.go:243 +0xb0
main.main()
/home/ajays/db/orm.go:13 +0x3a2

goroutine 19 [finalizer wait]:
runtime.park(0x413780, 0x8486a8, 0x847189)
/usr/share/go/src/pkg/runtime/proc.c:1369 +0x89
runtime.parkunlock(0x8486a8, 0x847189)
/usr/share/go/src/pkg/runtime/proc.c:1385 +0x3b
runfinq()
/usr/share/go/src/pkg/runtime/mgc0.c:2644 +0xcf
runtime.goexit()
/usr/share/go/src/pkg/runtime/proc.c:1445

goroutine 20 [chan receive]:
database/sql.(*DB).connectionOpener(0xc208048180)
/usr/share/go/src/pkg/database/sql/sql.go:583 +0x48
created by database/sql.Open
/usr/share/go/src/pkg/database/sql/sql.go:442 +0x27c

goroutine 21 [runnable]:
github.com/jinzhu/gorm.(*DB).print(0xc208040d10, 0xc2080475a0, 0x2, 0x2)
/usr/share/go/bin/src/github.com/jinzhu/gorm/main_private.go:36
created by github.com/jinzhu/gorm.(*DB).err
/usr/share/go/bin/src/github.com/jinzhu/gorm/main_private.go:26 +0x1a3

On Wed, Mar 4, 2015 at 12:32 PM, Jinzhu <notifications@github.com> wrote:

> Closed #400 <https://github.com/jinzhu/gorm/issues/400>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jinzhu/gorm/issues/400#event-245788690>.
>



-- 
*--*
*Cheers!*
*Ajay.S*

**Comment by ajays20078** [comment]
@jinzhu 

How do i get the IDs in the above query, i need the IDs whose status have been set to allocated.

**Comment by ajays20078** [comment]
@jinzhu 

The command you mentioned translates to - 

UPDATE `resources` SET `status` = 'allotted'  WHERE (status = 'available') LIMIT 1

Here you are not recording what IDs got affected, for that you need to do a select.

So you are essentially skipping "select on update" and thus loosing out on IDs that got affected, which is the data which needs to be returned(which resource got allotted)

**Comment by jinzhu** [comment]
You can't get ids with above query, basically you want to have a lock when update resource's state, yes?

You could do that with `sync` package or use raw sql https://github.com/jinzhu/gorm#raw-sql to do the work

**Comment by ajays20078** [comment]
Whats the point of using a ORM if i have to execute a RAW sql?
And sync will be app layer lock, not a DB layer, so your rows/table can still be updated.
This orm will be useless for transactions if you do not support locking.

**Comment by jinzhu** [comment]
Because even use raw sql with gorm to do that is still pretty easy

```go
var ids []int
db.Row("select id from resources for update limit 5").Scan(&ids)
db.Where("Id in (?)", ids).Update(&Resource{State: "allotted"})
```

I don't think ORM could handle 100% cases, just use it to do those things that it good at.

Your case is not a common thing, and I can't imagine a good solution to improve gorm to make it fit your case well. (If you have ideas please suggest it)

**Comment by ajays20078** [comment]
@jinzhu 
 

Your way of updating resource state does NOT work for concurrent request.

Consider the following sequence of execution

Request 1: var ids []int
Request 2: var ids []int
Request 1: db.Row("select id from resources for update limit 5").Scan(&ids) <- gets 5 resources
Request 2: db.Row("select id from resources for update limit 5").Scan(&ids) <- gets same 5 resource
Request1:db.Where("Id in (?)", ids).Update(&Resource{State: "allotted"}) <- updates 5 rows
Request2:db.Where("Id in (?)", ids).Update(&Resource{State: "allotted"}) <- updates same 5 rows.

You end up updating same 5 rows.

One way of implementing locks in ORM is,you can check before update, if the ID which you have read in memory is same as the ID on Disk.

**Comment by jinzhu** [comment]
'select for update' need to be run in a transaction, and it will prevent the select in second session until you commit current transaction, so you need to run above code in a transaction like this:

```go
tx := db.Begin()
var ids []int
tx.Row("select id from resources limit 5 for update").Scan(&ids)
tx.Where("Id in (?)", ids).Update(&Resource{State: "allotted"})
tx.Commit()
```

**Comment by ajays20078** [comment]
This is exactly what i was trying to explain. We need the same behavior through ORM abstraction and not through raw query "select for update".

**Comment by gdm85** [comment]
I think the above should be at least mentioned in documentation, as it is a very common use-case nowadays (who doesn't want concurrency, or does not implement it without considering race conditions?)

@ajays20078 shall we make a fork/pull request for this? I think it's very important to support it in the ORM (or stop using ORMs altogether)

**Comment by zeemetz** [comment]
its almost 2 year now.

Do we have the locking function now? because, concurrent is one of golang most wanted feature. let assume  we have lastbalance that we want to update concurrently, without locking we wont have correct last balance in the end of request. 

example:
update account set last_balance = last_balance - nominal where id=?

when this query executed by 2 concurrent request, they both would get last_balance before update with same value

before update:
last_balance = 10.000

request 1 nominal = 1.000
request 2 nominal = 2.000

expected last_balance = 7.000

actual last_balance without locking = 8.000 or 9.000 depend on which request saved first, because:
10.000-1.000
10.000-2.000

If there are any idea to make this happen please share.
the only solution now i used is handling the balance deduction with database trigger

Thank You

**Comment by zhuharev** [comment]
@jinzhu,  can you reopen this issue? 

I think it should be like this https://github.com/go-gorp/gorp#optimistic-locking

**Comment by iorlas** [comment]
Guys, hey guys. It was solved years ago: #768

Look at commit c9dfd80959ef31311810835abeaaf5d07131089c 

Gonna test it some time today
